internal/jsonrpc2/jsonrpc2.go

```diff
@@ -0,0 +1,282 @@
+
+type Conn struct {
+	listenErr            chan error
+	currentId            int64
+	opened               bool
+	stream               MessageStream
+	requestsHandlers     map[string]requestHandler
+	notificationHandlers map[string]notificationHandler
+	pendingRequestMu     sync.RWMutex
+	pendingRequests      map[ID](chan Response)
+}
+
+type Handler struct {
+	register func(*Conn)
+}
+
+type HandlingStrategy uint8
+
+const (
+	SyncHandling HandlingStrategy = iota
+	AsyncHandling
+)
+
+// Create a request handler for the given method
+//
+// The handler will be called asynchronously. Will panic if not able to marshal the handler's return value
+func NewRequestHandler[Params any](method string, strategy HandlingStrategy, handler func(params Params, conn *Conn) any) Handler {
+
+	return Handler{
+		register: func(conn *Conn) {
+			requestHandler := func(id ID, raw json.RawMessage) {
+				var payload Params
+				if raw != nil {
+					err := json.Unmarshal([]byte(raw), &payload)
+					if err != nil {
+						conn.stream.WriteMessage(Response{
+							ID:    id,
+							Error: &ErrInvalidParams,
+						})
+						return
+					}
+				}
+
+				out := handler(payload, conn)
+
+				bytes, err := json.Marshal(out)
+				if err != nil {
+					panic(err)
+				}
+
+				conn.stream.WriteMessage(Response{
+					ID:     id,
+					Result: bytes,
+				})
+			}
+
+			switch strategy {
+			case SyncHandling:
+				conn.requestsHandlers[method] = requestHandler
+
+			case AsyncHandling:
+				conn.requestsHandlers[method] = func(id ID, raw json.RawMessage) {
+					go requestHandler(id, raw)
+				}
+
+			default:
+				panic("bad argument: Invalid handling strategy")
+			}
+		},
+	}
+}
+
+// Create a notification handler for the given method
+//
+// The handler will be called asynchronously
+func NewNotificationHandler[Params any](method string, strategy HandlingStrategy, handler func(params Params, conn *Conn)) Handler {
+	return Handler{
+		register: func(conn *Conn) {
+			notificationHandler := func(raw json.RawMessage) {
+				var payload Params
+				err := json.Unmarshal([]byte(raw), &payload)
+				if err != nil {
+					// as per the json-rpc2 specs (https://www.jsonrpc.org/specification#notification),
+					// even if there are error in the notification payload, we won't return a response
+					return
+				}
+
+				handler(payload, conn)
+			}
+
+			switch strategy {
+			case SyncHandling:
+				conn.notificationHandlers[method] = notificationHandler
+
+			case AsyncHandling:
+				conn.notificationHandlers[method] = func(raw json.RawMessage) {
+					go notificationHandler(raw)
+				}
+
+			default:
+				panic("bad argument: Invalid handling strategy")
+			}
+
+		},
+	}
+}
+
+// Starts listening asynchronously to the MessageStream and returns the connection.
+//
+// By default, the server will try write concurrently to the ObjectStream
+func NewConn(objStream MessageStream, handlers ...Handler) *Conn {
+	conn := Conn{
+		listenErr:            make(chan error),
```

Comment:
If no listener are listening this channel, your loop will block.
You can create a chan of size 1 to avoid that.
```
make(chan error, 1)
```